ws-server/src/server.rs

```diff
@@ -263,11 +233,12 @@ async fn background_task(
 
 	// Buffer for incoming data.
 	let mut data = Vec::with_capacity(100);
+	let mut method_executors = FutureDriver::default();
 
 	while !shutdown.is_closed() {
 		data.clear();
 
-		receiver.receive_data(&mut data).await?;
+		method_executors.select_with(receiver.receive_data(&mut data)).await?;
```

Comment:
If a client has a WS connection open and so `receive_data` here is waiting for incoming data from them, we won't loop around and notice that we've been `shutdown.is_closed()` is true and we should finish.

So, will shutdown not complete until all open connections from clients have been voluntarily closed?

Should we `select` on shutdown + receive_data here so that we can see as soon as we've shutdown and can end this loop (which I think will also end the task spawned above gracefully)? (receive_data isn't cancel safe, but I don't think it matters if we're shutting down anyway.. at least I made the same assumption in telemetry!)